src/samplers/ldsampler.cpp

```diff
@@ -0,0 +1,135 @@
+             this sampler doesn't add additional jittering.
+.. subfigend::
+   :label: fig-ldsampler-pattern
+
+.. subfigstart::
+.. subfigure:: ../../resources/data/docs/images/sampler/ldsampler_1024_samples_dim_32.svg
+   :caption: 1024 samples projected onto the 32th and 33th dimensions, which look almost identical.
+             However, note that the points have been scrambled to reduce correlations between
+             dimensions.
+.. subfigure:: ../../resources/data/docs/images/sampler/ldsampler_64_samples_and_proj_dim_32.svg
+   :caption: A projection of the first 64 samples onto the 32th and 33th dimensions.
+.. subfigend::
+   :label: fig-ldsampler-pattern_dim32
+
+ */
+
+template <typename Float, typename Spectrum>
+class LowDiscrepancySampler  final : public Sampler<Float, Spectrum> {
+public:
+    MTS_IMPORT_BASE(Sampler, m_sample_count, m_base_seed,
+                    m_samples_per_wavefront, m_dimension_index,
+                    current_sample_index, compute_per_sequence_seed)
+    MTS_IMPORT_TYPES()
+
+    LowDiscrepancySampler (const Properties &props = Properties()) : Base(props) {
+        // Make sure sample_count is power of two and square (e.g. 4, 16, 64, 256, 1024, ...)
+        ScalarUInt32 res = 2;
+        while (sqr(res) < m_sample_count)
+            res = math::round_to_power_of_two(++res);
+
+        if (m_sample_count != sqr(res))
+            Log(Warn, "Sample count should be square and power of two, rounding to %i", sqr(res));
+
+        m_sample_count = sqr(res);
+    }
+
+    ref<Sampler<Float, Spectrum>> clone() override {
+        LowDiscrepancySampler  *sampler  = new LowDiscrepancySampler ();
+        sampler->m_sample_count          = m_sample_count;
+        sampler->m_samples_per_wavefront = m_samples_per_wavefront;
+        sampler->m_base_seed             = m_base_seed;
+        return sampler;
+    }
+
+    void seed(uint64_t seed_offset, size_t wavefront_size) override {
+        Base::seed(seed_offset, wavefront_size);
+        m_scramble_seed = compute_per_sequence_seed(seed_offset);
+    }
+
+    Float next_1d(Mask active = true) override {
+        ENOKI_MARK_USED(active);
+
+        UInt32 sample_indices = current_sample_index();
+        UInt32 perm_seed = m_scramble_seed + m_dimension_index++;
+
+        UInt32 i = permute(sample_indices, m_sample_count, perm_seed);
+        UInt32 scramble = sample_tea_32(m_scramble_seed, UInt32(0x48bc48eb));
+
+        return radical_inverse_2(i, scramble);
+    }
+
+    Point2f next_2d(Mask active = true) override {
+        ENOKI_MARK_USED(active);
+
+        UInt32 sample_indices = current_sample_index();
+        UInt32 perm_seed = m_scramble_seed + m_dimension_index++;
+
+        UInt32 i = permute(sample_indices, m_sample_count, perm_seed);
+        UInt32 scramble_x = sample_tea_32(m_scramble_seed, UInt32(0x18bc21eb));
+        UInt32 scramble_y = sample_tea_32(m_scramble_seed, UInt32(0x58633e2d));
```

Comment:
Ditto, also where are these constants coming from?

Reply:
Those are arbitrary constants. We use them here to get different scramble values along the x and y axis from the same scramble seed.